app/models/carriage.rb

```diff
@@ -1,9 +1,11 @@
 class Carriage < ApplicationRecord
-  validates :bottom_places, presence: true
-  validates :upper_places, presence: true
-  validates :carriage_type, presence: true
-  validates :train, presence: true
-
-  belongs_to :carriage_type
   belongs_to :train
+
+  before_validation :set_number
+
+  protected
+
+  def set_number
+    self.number ||= train.carriages.size + 1 if train
```

Comment:
I think better use https://zaiste.net/rails_activerecord_updating_attributes_object_fields/ instead of _self._